### PR TITLE
Cloud - Trash

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        backupHelper = OABackupHelper.sharedInstance()
+        settingsHelper = OANetworkSettingsHelper.sharedInstance()
+    }
+
+    override func registerNotifications() {
+        addNotification(NSNotification.Name(kBackupProgressUpdateNotification), selector: #selector(onBackupProgressUpdate(notification:)))
+        addNotification(NSNotification.Name(kBackupSyncFinishedNotification), selector: #selector(onBackupSyncFinished(notification:)))
+    }
+
+    deinit {
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:56]: move under viewDidLoad

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+            }
+        }
+        return items
+    }
+
+    private func collectTrashGroups() -> [String: TrashGroup] {
+        var groups: [String: TrashGroup] = [:]
+        var items: [TrashItem] = collectTrashItems()
+        if !items.isEmpty {
+            items.sort(by: { i1, i2 in
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:142]: $0, $1

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+                i1.time > i2.time
+            })
+            
+            for item in items {
+                let time: Int = item.time
+                let dateFormatter = DateFormatter()
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:148]: crate lazy var for dateFormatter

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                    progressView.iconPath = tickPath(progressView)
+                    progressView.stopSpinProgressBackgroundLayer()
+                }
+                cell.accessoryType = .none
+                cell.accessoryView = nil
+            }
+        }
+    }
+
+    private func tickPath(_ progressView: FFCircularProgressView) -> UIBezierPath {
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:315]: OAResourcesUIHelper ?

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+        }
+    }
+
+    func downloadItem(_ trashItem: TrashItem) {
+        guard trashItem.settingsItem != nil else {
+            Self.logger.error("Failed to download item: \(String(describing: trashItem.oldFile.name)), SettingsItem is null")
+            return
+        }
+        if backupHelper?.backup.localFiles[trashItem.oldFile.getTypeNamePath()] != nil {
+            downloadItem(trashItem, shouldReplace: false) // TODO: - android uses FileExistBottomSheet here
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:401]: ?

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+            if !deleteAll, let indexPath = self.selectedIndexPath {
+                self.updateRowProgress(.started, indexPath: indexPath)
+            } else if self.tableView.refreshControl?.isRefreshing == false {
+                self.tableView.refreshControl?.beginRefreshing()
+            }
+        }
+    }
+
+    func onFileDeleteProgress(_ deleteAll: Bool) {
+        if !deleteAll, let indexPath = self.selectedIndexPath {
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:435]: if let selectedIndexPath  {}

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+                self.updateRowProgress(.inProgress, indexPath: indexPath)
+            }
+        }
+    }
+
+    func onFilesDeleteDone(_ message: String, errors: [OARemoteFile: String], deleteAll: Bool) {
+        let hasErrors = !errors.isEmpty
+        if hasErrors || !message.isEmpty {
+            DispatchQueue.main.async {
+                if !deleteAll, let indexPath = self.selectedIndexPath {
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:446]: if let selectedIndexPath

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+        if !deleteAll, let indexPath = self.selectedIndexPath {
+            DispatchQueue.main.async {
+                self.updateRowProgress(.inProgress, indexPath: indexPath)
+            }
+        }
+    }
+
+    func onFilesDeleteDone(_ message: String, errors: [OARemoteFile: String], deleteAll: Bool) {
+        let hasErrors = !errors.isEmpty
+        if hasErrors || !message.isEmpty {
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:444]: if !errors.isEmpty|| !message.isEmpty {

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+        }
+        backupHelper?.prepareBackup()
+    }
+}
+
+private final class TrashDeletionListener: NSObject, OAOnDeleteFilesListener {
+
+    weak var delegate: CloudTrashDeletionDelegate?
+    private let messageDone: String
+    private var deleteAll: Bool = false
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:473]: remove bool

File: Sources/Controllers/Cloud/OACloudBackupViewController.mm [hunk 10]
```diff
@@ -692,13 +740,25 @@ - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath
     {
         statusBackupViewController = [[OAStatusBackupViewController alloc] initWithType:EOARecentChangesConflicts];
     }
+    else if ([item.key isEqualToString:@"onTrashPressed"])
+    {
+        if ([OAIAPHelper isOsmAndProAvailable])
+        {
+            OACloudTrashViewController *cloudTrashController = [[OACloudTrashViewController alloc] init];
```
Comment (tigrim) [inline Sources/Controllers/Cloud/OACloudBackupViewController.mm:747]: don't use temp variable

File: Sources/Helpers/OAOsmAndFormatter.mm [hunk 11]
```diff
@@ -100,6 +100,14 @@ + (NSString *) getFormattedDate:(NSTimeInterval)time
     NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
     [formatter setFormatterBehavior:NSDateFormatterBehavior10_4];
     [formatter setDateStyle:NSDateFormatterMediumStyle];
+    return [formatter stringFromDate:[NSDate dateWithTimeIntervalSince1970:time]];
+}
+
++ (NSString *)getFormattedDateTime:(NSTimeInterval)time
```
Comment (tigrim) [inline Sources/Helpers/OAOsmAndFormatter.mm:106]: Can you use a more understandable naming?

File: Sources/Backup/BackupUiUtils.swift [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+    static func getItemName(_ item: OASettingsItem) -> String {
+        var name: String
+        if let profileItem = item as? OAProfileSettingsItem {
+            name = profileItem.appMode.toHumanString()
+        } else {
+            name = item.getPublicName()
+            if let fileItem = item as? OAFileSettingsItem {
+                let subtype = fileItem.subtype
+                if subtype == .subtypeVoiceTTS {
```
Comment (tigrim) [inline Sources/Backup/BackupUiUtils.swift:26]: use switch

File: Sources/Backup/BackupUiUtils.swift [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        return type?.icon
+    }
+
+    static func getLastBackupTimeDescription(_ def: String) -> String {
+        let lastUploadedTime = OAAppSettings.sharedManager().backupLastUploadedTime.get()
+        return getFormattedPassedTime(time: lastUploadedTime, def: def, showTime: false)
+    }
+
+    static func getFormattedPassedTime(time: Int, def: String, showTime: Bool) -> String {
+        if time > 0 {
```
Comment (tigrim) [inline Sources/Backup/BackupUiUtils.swift:52]: guard

File: Sources/Backup/BackupUiUtils.swift [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    static func generateTimeString(summary: String, time: Int) -> String {
+        String(format: localizedString("ltr_or_rtl_combine_via_colon"), summary, getTimeString(time))
+    }
+
+    static func getTimeString(_ time: Int) -> String {
+        formatPassedTime(time: time, longPattern: "d MMM yyyy, HH:mm:ss", shortPattern: "HH:mm:ss", def: localizedString("shared_string_never"))
+    }
+
+    static func formatPassedTime(time: Int, longPattern: String, shortPattern: String, def: String) -> String {
+        if time > 0 {
```
Comment (tigrim) [inline Sources/Backup/BackupUiUtils.swift:73]: guard

File: Sources/Controllers/Cloud/CloudTrashItemMenuViewController.swift [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        deleteRow.secondaryIconName = "ic_custom_trash_outlined"
+        deleteRow.secondaryIconTintColor = UIColor.iconColorDisruptive
+    }
+
+    override func getRow(_ indexPath: IndexPath) -> UITableViewCell? {
+        let item = tableData.item(for: indexPath)
+        if item.cellType == OASimpleTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OASimpleTableViewCell.getIdentifier()) as? OASimpleTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OASimpleTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashItemMenuViewController.swift:95]: Do not use this code. It's a deprecated approach. Instead, use tableView.register() in all cases. Example: tableView.register(UINib(nibName: OASimpleTableViewCell.getIdentifier(), bundle: nil), forCellReuseIdentifier: OASimpleTableViewCell.getIdentifier())

File: Sources/Controllers/Cloud/CloudTrashItemMenuViewController.swift [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+                        iconColor = UIColor.iconColorDefault
+                    }
+                    cell.leftIconView.tintColor = iconColor
+                }
+            }
+            return cell
+        } else if item.cellType == OARightIconTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OARightIconTableViewCell.getIdentifier()) as? OARightIconTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OARightIconTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashItemMenuViewController.swift:120]: See above

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 17]
```diff
@@ -0,0 +1,499 @@
...
+        }
+        return groups
+    }
+
+    override func getRow(_ indexPath: IndexPath) -> UITableViewCell? {
+        let item: OATableRowData = tableData.item(for: indexPath)
+        if item.cellType == OASimpleTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OASimpleTableViewCell.getIdentifier()) as? OASimpleTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OASimpleTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:168]: See above

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 18]
```diff
@@ -0,0 +1,499 @@
...
+                        iconColor = UIColor.iconColorDefault
+                    }
+                    cell.leftIconView.tintColor = iconColor
+                }
+            }
+            return cell
+        } else if item.cellType == OALargeImageTitleDescrTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OALargeImageTitleDescrTableViewCell.getIdentifier()) as? OALargeImageTitleDescrTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OALargeImageTitleDescrTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:193]: See above

File: Sources/Controllers/Cloud/CloudTrashViewController.swift [hunk 19]
```diff
@@ -0,0 +1,499 @@
...
+                cell?.showButton(false)
+            }
+            if let cell {
+                cell.titleLabel?.text = item.title
+                cell.titleLabel?.accessibilityLabel = item.title
+                cell.descriptionLabel?.text = item.descr
+                cell.descriptionLabel?.accessibilityLabel = item.descr
+                cell.cellImageView?.tintColor = item.iconTintColor
+                if let icon = item.iconName {
+                    cell.cellImageView?.image = OAUtilities.resize(UIImage(named: icon), newSize: CGSize(width: 60.0, height: 60.0)).withRenderingMode(.alwaysTemplate)
```
Comment (tigrim) [inline Sources/Controllers/Cloud/CloudTrashViewController.swift:205]: We always need to resize, maybe cache already resized images?

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/OAOsmAndFormatter.mm [hunk 20]
```diff
@@ -694,4 +702,27 @@ + (NSString *) getFormattedDuration:(NSTimeInterval)seconds fullForm:(BOOL) full
       }
 }
 
++ (NSString *)getFormattedDuration:(NSTimeInterval)seconds
```
Comment (alex-osm) [inline Sources/Helpers/OAOsmAndFormatter.mm:705]: Same name with previous method. Hard to understand what is difference.

**Pull request merged**